lib/block_view_atomic_txns.go

```diff
@@ -0,0 +1,557 @@
+		if err != nil {
+			return errors.Wrap(err, "_verifyAtomicTxnsChain")
+		}
+		atomicHashes = append(atomicHashes, innerTxnAtomicHash.ToBytes())
+	}
+
+	// Validate the chain sequence specified.
+	for ii, innerTxn := range txnMeta.Txns {
+		// Check the next transaction.
+		nextIndex := (ii + 1) % len(txnMeta.Txns)
+		if !bytes.Equal(
+			innerTxn.ExtraData[NextAtomicTxnPreHash],
+			atomicHashes[nextIndex]) {
+			return RuleErrorAtomicTxnsHasBrokenChain
+		}
+
+		// Check the previous transaction
+		prevIndex := (ii - 1 + len(txnMeta.Txns)) % len(txnMeta.Txns)
+		if !bytes.Equal(
+			innerTxn.ExtraData[PreviousAtomicTxnPreHash],
+			atomicHashes[prevIndex]) {
+			return RuleErrorAtomicTxnsHasBrokenChain
+		}
+	}
+	return nil
+}
+
+func (bav *UtxoView) _disconnectAtomicTxnsWrapper(
+	operationType OperationType,
+	currentTxn *MsgDeSoTxn,
+	txnHash *BlockHash,
+	utxoOpsForTxn []*UtxoOperation,
+	blockHeight uint32,
+) error {
+	if len(utxoOpsForTxn) == 0 {
+		return fmt.Errorf("_disconnectAtomicTxnsWrapper: utxoOperations are missing")
+	}
+	operationIndex := len(utxoOpsForTxn) - 1
+
+	// Verify the last operation as being of type OperationTypeAtomicTxns.
+	if utxoOpsForTxn[operationIndex].Type != OperationTypeAtomicTxnsWrapper {
+		return fmt.Errorf("_disconnectAtomicTxnsWrapper: Trying to revert "+
+			"OperationTypeAtomicTxns but found type %v", utxoOpsForTxn[operationIndex].Type)
+	}
+
+	// Gather the transaction metadata so we know the internal transactions.
+	txMeta := currentTxn.TxnMeta.(*AtomicTxnsWrapperMetadata)
+
+	// Sanity check the AtomicTxns operation exists.
+	operationData := utxoOpsForTxn[operationIndex]
+	if operationData.AtomicTxnsInnerUtxoOps == nil ||
+		len(operationData.AtomicTxnsInnerUtxoOps) != len(txMeta.Txns) {
+		return fmt.Errorf("_disconnectAtomicTxnsWrapper: Trying to revert OperationTypeAtomicTxns " +
+			"but found nil or mistmatched number of UtxoOps for inner transactions")
+	}
+
+	// Disconnect the internal transactions in reverse.
+	for ii := len(txMeta.Txns) - 1; ii >= 0; ii-- {
+		innerTxn := txMeta.Txns[ii]
+
+		if err := bav.DisconnectTransaction(
+			innerTxn,
+			innerTxn.Hash(),
+			operationData.AtomicTxnsInnerUtxoOps[ii],
+			blockHeight); err != nil {
+			return errors.Wrapf(err, "_disconnectAtomicTxnsWrapper")
+		}
+	}
+
+	return nil
+}
+
+//
+// TYPES: AtomicTxnsWrapperTxindexMetadata
+//
+
+type AtomicTxnsWrapperTxindexMetadata struct {
+	InnerTxnsTransactionMetadata []*TransactionMetadata
+}
+
+func (txindexMetadata *AtomicTxnsWrapperTxindexMetadata) RawEncodeWithoutMetadata(
+	blockHeight uint64,
+	skipMetadata ...bool,
+) []byte {
+	var data []byte
+	data = append(data, UintToBuf(uint64(len(txindexMetadata.InnerTxnsTransactionMetadata)))...)
+	for _, innerMetadata := range txindexMetadata.InnerTxnsTransactionMetadata {
+		txnBytes := innerMetadata.RawEncodeWithoutMetadata(blockHeight, skipMetadata...)
+		data = append(data, UintToBuf(uint64(len(txnBytes)))...)
```

Comment:
```suggestion
```
we don't need the number of bytes - the RawEncodeWithoutMetadata and RawDecodeWithoutMetadata functions will read the exact number of bytes required.